### PR TITLE
Add FenViewer IA

### PR DESCRIPTION
**Pull request by rouzbeh**
**What does your Instant Answer do?**
This instant answer parses a FEN formatted chess position and displays it on a chessboard.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
It's a very quick way to visualise a chess position.

**What is the data source for your Instant Answer? (Provide a link if possible)**
There is no data source.

**Why did you choose this data source?**
NA

**Are there any other alternative (better) data sources?**
NA

**What are some example queries that trigger this Instant Answer?**
FEN rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Chess players

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
No

**Which existing Instant Answers will this one supercede/overlap with?**
NA

**Are you having any problems? Do you need our help with anything?**
No

**Where did you hear about DuckDuckHack? (For first time contributors)**
On the DuckDuckGo main page.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
http://imgur.com/XzHecj0

## Checklist
Please place an 'X' where appropriate.

```
- [X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
- [X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
- [X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
- [X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
- [X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        - [X] Google Chrome
        - [] Firefox
        - [] Opera
        - [] IE 10
        - [X] IE 11

    - Windows 7
        - [] Google Chrome
        - [] Firefox
        - [] Opera
        - [] IE 8
        - [] IE 9
        - [] IE 10

    - Windows XP
        - [] IE 7
        - [] IE 8

    - Mac OSX
        - [] Google Chrome
        - [] Firefox
        - [] Opera
        - [] Safari

    - iOS (iPhone)
        - [] Safari Mobile
        - [] Google Chrome
        - [] Opera

    - iOS (iPad)
        - [] Safari Mobile
        - [] Google Chrome
        - [] Opera

    - Android
        - [X] Firefox
        - [] Native Browser
        - [X] Google Chrome
        - [] Opera
```

### CODE REVIEW

File: lib/DDG/Goodie/FenViewer.pm [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+# that the given position is actually legal.
+
+use DDG::Goodie;
+use Scalar::Util qw(looks_like_number);
+use Try::Tiny;
+
+zci answer_type => "fen_viewer";
+zci is_cached   => 1;
+
+# Metadata.  See https://duck.co/duckduckhack/metadata for help in filling out this section.
```
Comment (mintsoft) [inline lib/DDG/Goodie/FenViewer.pm]: You can nuke these instructional comments

File: t/FenViewer.t [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
...s="empty"></a></td><td id="E3"><a href="#" class="empty"></a></td><td id="F3"><a href="#" class="empty"></a></td><td id="G3"><a href="#" class="empty"></a></td><td id="H3"><a href="#" class="empty"></a></td></tr><tr><td id="A2"><a href="#" class="white pawn">&#9817;</a></td><td id="B2"><a href="#" class="white pawn">&#9817;</a></td><td id="C2"><a href="#" class="white pawn">&#9817;</a></td><td id="D2"><a href="#" class="white pawn">&#9817;</a></td><td id="E2"><a href="#" class="white pawn">&#9817;</a></td><td id="F2"><a href="#" class="white pawn">&#9817;</a></td><td id="G2"><a href="#" class="white pawn">&#9817;</a></td><td id="H2"><a href="#" class="white pawn">&#9817;</a></td></tr><tr><td id="A1"><a href="#" class="white rook">&#9814;</a></td><td id="B1"><a href="#" class="white knight">&#9816;</a></td><td id="C1"><a href="#" class="white bishop">&#9815;</a></td><td id="D1"><a href="#" class="white queen">&#9813;</a></td><td id="E1"><a href="#" class="white king">&#9812;</a></td><td id="F1"><a href="#" class="white bishop">&#9815;</a></td><td id="G1"><a href="#" class="white knight">&#9816;</a></td><td id="H1"><a href="#" class="white rook">&#9814;</a></td></tr></table></div>'),
```
Comment (mintsoft) [inline t/FenViewer.t]: Given the amount of parsing in the IA can you provide some "real" test cases for non-starting positions etc?

File: share/goodie/fen_viewer/fen_viewer.css [hunk 3]
```diff
@@ -0,0 +1,35 @@
+/*
+    Document   : zeroclickinfo-goodies/share/goodie/fen_viewer/fen_viewer.css
```
Comment (mintsoft) [inline share/goodie/fen_viewer/fen_viewer.css]: This comment can also go :+1:

File: lib/DDG/Goodie/FenViewer.pm [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    return unless $query;
+    my (@pos) = parse_position($query);
+    my ($html_out) = '';
+    try {
+        $html_out = draw_chessboard(@pos);
+    }
+    catch {
+        return;
+    };
+    return 'Chessboard', html => $html_out;
```
Comment (mintsoft) [inline lib/DDG/Goodie/FenViewer.pm]: It'd be nice if the text output was an ascii representation of the board; similiar to how Sudoku does it: https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Sudoku.pm

File: lib/DDG/Goodie/FenViewer.pm [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    return $html_chessboard;
+}
+
+# Handle statement
+handle remainder => sub {
+
+    # optional - regex guard
+    # return unless qr/^\w+/;
+    my ($query) = $_;
+    return unless $query;
```
Comment (mintsoft) [inline lib/DDG/Goodie/FenViewer.pm]: Is it worth actually guarding this with regex? How well behaved is the parsing with invalid input?

File: t/FenViewer.t [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    # Try to include some examples of queries on which it might
+    # appear that your answer will trigger, but does not.
+    'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1' => undef,
```
Comment (mintsoft) [inline t/FenViewer.t]: Can you also add some queries like "fen" that should return `undef`

**Comment by mintsoft** [comment]
@rouzbeh This looks really cool! I could have sworn we already had this or similiar implemented somewhere, however I now think I might have been getting confused with the Chess960 IA: https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Chess960.pm

Can you add a screenshot of the IA?

**Comment by rouzbeh** [comment]
@mintsoft Thanks a lot for the review and your suggestions.

I have added ASCII output, removed useless comments, and made the parsing more robust by checking for more cases of error. There are now also more tests, including two positions from the world championship match. There was already a screenshot in the pull request. Here it is again : http://imgur.com/XzHecj0

One advantage of this IA, compared to Chess960, is that it does everything in HTML. No querying for images.

File: lib/DDG/Goodie/FenViewer.pm [hunk 7]
```diff
@@ -0,0 +1,160 @@
+package DDG::Goodie::FenViewer;
+# ABSTRACT: This instant answer parses a chess position in the Forsyth-Edwards notation, 
+# and draws a chessboard on screen representing that position. The current version only 
+# parses the current board position (the first field in the FEN format) and does not check
+# that the given position is actually legal.
+
+use DDG::Goodie;
```
Comment (mintsoft) [inline lib/DDG/Goodie/FenViewer.pm]: Can you add `use strict;` here? It'll force github to recognise it as perl 5

**Comment by mintsoft** [comment]
@abeyang any design feedback? http://imgur.com/XzHecj0

**Comment by rouzbeh** [comment]
Hi

Any news on the design review ? Once this is done, I can update the Chess960 to use the same visual output instead of an image.

Cheers,
Ali

**Comment by moollaza** [comment]
@rouzbeh sorry for the delay, we'll get around to this shortly!

**Comment by abeyang** [comment]
Hey @rouzbeh , this looks pretty neat! Can we clean up the visuals, just a tad:

```css
.zci--answer .zci--fenviewer a {
  color: #333;
  font-size: 27px;
  line-height: 1.5;
  text-shadow: none; /* remove text-shadow altogether */
}

.zci--answer .zci--fenviewer .chess_board tr:nth-child(odd) td:nth-child(even), .zci--answer .zci--fenviewer .chess_board tr:nth-child(even) td:nth-child(odd) {
  background-color: #ddd;
  box-shadow: none; /* remove box-shadow altogether */
}

.zci--answer .zci--fenviewer .chess_board {
  border: 2px solid #333;
}
```

In the end, it should look something like this:
screen shot 2015-03-25 at 11 03 27 am

**Comment by rouzbeh** [comment]
This is how it looks like now : 

image

**Comment by abeyang** [comment]
Hi @rouzbeh , almost there! You're still missing the `line-height: 1.5;` for `.zci--answer .zci--fenviewer a`. Other than that, it looks good!

**Comment by rouzbeh** [comment]
@abeyang Oops, sorry about that. It should be good now.

**Comment by abeyang** [comment]
Looks good! :+1:

**Comment by moollaza** [comment]
LGTM :+1:

**Comment by moollaza** [comment]
@rouzbeh congrats! It's live: https://duckduckgo.com/?q=FEN+rnbqkbnr%2Fpppppppp%2F8%2F8%2F8%2F8%2FPPPPPPPP%2FRNBQKBNR+w+KQkq+-+0+1&ia=answer

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

**Comment by rouzbeh** [comment]
@moollaza Hooray :)

**Pull request merged**